File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/brand/InfoController.java
package com.goexp.galgame.gui.view.brand;

import com.goexp.galgame.gui.model.Brand;
import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.task.game.GameSearchTask;
import com.goexp.galgame.gui.util.DefaultController;
import com.goexp.galgame.gui.util.TaskService;
import com.goexp.galgame.gui.view.game.listview.DataViewController;
import javafx.beans.value.ChangeListener;
import javafx.collections.ObservableList;
import javafx.collections.transformation.FilteredList;
import javafx.concurrent.Service;
import javafx.fxml.FXML;

public class InfoController extends DefaultController {

    /**
     * UI Com
     */

    @FXML
    public TitlePartController titleController;

    @FXML
    public DataViewController dataViewController;

    private FilteredList<Game> filteredGames;


    /***
     * Biz
     */

    private Brand brand = new Brand();

    /**
     * Service
     */


    private Service<ObservableList<Game>> gameByBrand = new TaskService<>(() -> new GameSearchTask.ByBrand(brand.id));


    /**
     * Event
     */


    protected void initialize() {


        ChangeListener<Throwable> exceptionHandler = (observable, oldValue, newValue) -> {
            if (newValue != null)
                newValue.printStackTrace();
        };

        ChangeListener<ObservableList<Game>> searchHandler = (observable, oldValue, newValue) -> {

            if (newValue != null) {
                load(newValue);
            }

        };

        gameByBrand.valueProperty().addListener(searchHandler);
        gameByBrand.exceptionProperty().addListener(exceptionHandler);

        dataViewController.progessloading.visibleProperty().bind(gameByBrand.runningProperty());
        dataViewController.reloadProperty.addListener((observable, oldValue, newValue) -> {
            if (newValue != null && newValue) {
                gameByBrand.restart();
            }
        });

    }


    private void load(ObservableList<Game> games) {
        filteredGames = new FilteredList<>(games);
        dataViewController.setItems(filteredGames);
    }

    public void load(Brand brand) {

        logger.info("<Brand> {}", brand);
        this.brand = brand;

        titleController.init(brand);
        gameByBrand.restart();
    }

}

File: desktopclient/src/main/scala/com/goexp/galgame/gui/view/brand/InfoController.scala
package com.goexp.galgame.gui.view.brand

import com.goexp.galgame.gui.model.{Brand, Game}
import com.goexp.galgame.gui.task.game.search.ByBrand
import com.goexp.galgame.gui.view.game.explorer.ExplorerController
import com.goexp.ui.javafx.{DefaultController, TaskService}
import com.mongodb.client.model.Filters
import javafx.beans.property.SimpleObjectProperty
import javafx.collections.ObservableList
import javafx.collections.transformation.FilteredList
import javafx.fxml.FXML
import scalafx.Includes._
import scalafx.beans.binding.Bindings

import java.util.Objects

class InfoController extends DefaultController {

  @FXML var titleController: TitlePartController = _
  @FXML var dataViewController: ExplorerController = _

  final lazy val brand = new SimpleObjectProperty[Brand]

  private val gameByBrand = TaskService(new ByBrand(brand.value.id))


  override protected def eventBinding(): Unit = {
    gameByBrand.value.onChange((_, _, newValue) => {
      if (newValue != null) {
        val filteredGames = new FilteredList[Game](newValue)

        dataViewController.load(filteredGames)
      }
    })


    brand.onChange((_, _, newValue) => {
      if (newValue != null) {

        logger.info(s"Brand[${newValue.id}] ${brand.name} state:<${newValue.state}>")

        titleController.brand = newValue
        titleController.compName <== Bindings.createStringBinding(() => newValue.comp)
        titleController.brandName <== Bindings.createStringBinding(() => newValue.name)
        titleController.state <==> newValue.state

        dataViewController.initFilter = Filters.eq("brandId", newValue.id)
        gameByBrand.restart()
      }

    })
  }
}